### PR TITLE
[SPARK-30236][SQL][DOCS][2.4] Clarify date and time patterns supported in the related functions

### PR DESCRIPTION
**Pull request by johnhany97**
### What changes were proposed in this pull request?

Link to appropriate Java Class with list of date/time patterns supported

### Why are the changes needed?

Avoid confusion on the end-user's side of things, as seen in questions like [this](https://stackoverflow.com/questions/54496878/date-format-conversion-is-adding-1-year-to-the-border-dates) on StackOverflow

### Does this PR introduce any user-facing change?

Yes, Docs are updated.

### How was this patch tested?

Built docs:
image

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 1]
```diff
@@ -541,7 +541,7 @@ case class WeekOfYear(child: Expression) extends UnaryExpression with ImplicitCa
 
 // scalastyle:off line.size.limit
 @ExpressionDescription(
-  usage = "_FUNC_(timestamp, fmt) - Converts `timestamp` to a value of string in the format specified by the date format `fmt`.",
+  usage = "_FUNC_(timestamp, fmt) - Converts `timestamp` to a value of string in the format specified by the date format `fmt`. See `java.text.SimpleDateFormat` for valid date and time format patterns.",
```
Comment (seanli-rallyhealth) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: use multiple line string? e.g.
"""
what ever 
"""

Comment (johnhany97) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Sure thing

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 2]
```diff
@@ -541,7 +541,10 @@ case class WeekOfYear(child: Expression) extends UnaryExpression with ImplicitCa
 
 // scalastyle:off line.size.limit
 @ExpressionDescription(
-  usage = "_FUNC_(timestamp, fmt) - Converts `timestamp` to a value of string in the format specified by the date format `fmt`.",
+  usage = """
+    _FUNC_(timestamp, fmt) - Converts `timestamp` to a value of string in the format specified by the date format `fmt`.
+    See `java.text.SimpleDateFormat` for valid date and time format patterns.
+  """,
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Hi, @johnhany97 .
Please open a PR to master first.

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: If we bypass `master` branch, Apache Spark 3.0+ document will have the same issue.

**Comment by dongjoon-hyun** [comment]
Oh, I saw #26864 . Let me take a look there.

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by srowen** [comment]
Yeah, do we need a separate PR for 2.4? looks like this might cherry pick cleanly. Is it definitely valid for 2.4?

**Comment by johnhany97** [comment]
2.4 uses a different implementation, so I reference here `java.text.SimpleDateFormat`. I'll update this with whatever the other PR on master resolves in the end to.

**Comment by johnhany97** [comment]
To elaborate on above comment, check this line out https://github.com/apache/spark/blob/5c67df7b958d4e45da907da8d87cb274e36b2348/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala#L619

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
The same comments here. You can update this after your PR is finalized at `master` branch.

**Comment by dongjoon-hyun** [comment]
The PR on master branch is merged. Please update this PR accordingly if needed. Thanks!

**Comment by johnhany97** [comment]
@dongjoon-hyun any changes you'd like added here or is this good to merge?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 3]
```diff
@@ -583,7 +589,14 @@ case class DateFormatClass(left: Expression, right: Expression, timeZoneId: Opti
  * Deterministic version of [[UnixTimestamp]], must have at least one parameter.
  */
 @ExpressionDescription(
-  usage = "_FUNC_(expr[, pattern]) - Returns the UNIX timestamp of the given time.",
+  usage = "_FUNC_(timeExp[, format]) - Returns the UNIX timestamp of the given time.",
+  arguments = """
+    Arguments:
+      * timeExp - A date/timestamp or string which is returned as a UNIX timestamp.
+      * format - Date/time format pattern to follow. Ignored if `timeExp` is not a string.
+                 Default value is "uuuu-MM-dd HH:mm:ss". See `java.text.SimpleDateFormat`
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: This default value looks wrong. Could you check once more in `branch-2.4`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 4]
```diff
@@ -622,7 +635,14 @@ case class ToUnixTimestamp(
  * second parameter.
  */
 @ExpressionDescription(
-  usage = "_FUNC_([expr[, pattern]]) - Returns the UNIX timestamp of current or specified time.",
+  usage = "_FUNC_([timeExp[, format]]) - Returns the UNIX timestamp of current or specified time.",
+  arguments = """
+    Arguments:
+      * timeExp - A date/timestamp or string. If not provided, this defaults to current time.
+      * format - Date/time format pattern to follow. Ignored if `timeExp` is not a string.
+                 Default value is "uuuu-MM-dd HH:mm:ss". See `java.text.SimpleDateFormat`
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: ditto.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
The default format looks wrong in two places. Could you fix them again in order to match with `branch-2.4` code?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala (hunk 3 above)
Comment (johnhany97) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Good catch! Resolved.

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @johnhany97 .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @johnhany97 . Merged to branch-2.4.

**Pull request closed without merging**